### PR TITLE
fix(pipeline): pass terminal-aware discount to HordeActorCriticAgent.update (#2344)

### LINKED ISSUE #2344: AlbertaPipeline control_mode="horde_ac" drops `terminated`: the value head bootstraps and the actor trace survives every episode boundary
`AlbertaPipeline.update` documents a `terminated` argument ("Scalar termination
flag (`0.0` or `1.0`)") and validates it as a float32 scalar restricted to
`{0, 1}`. On the default `control_mode="sarsa"` path it is forwarded into
`step4_update` → `SARSAAgent.update`, where it skips bootstrapping
(`skip_bootstrap = (safe_terminated != 0) | (gamma_arr == 0.0)`,
`alberta_framework/core/sarsa.py:945`) and zeroes the eligibility traces
(`sarsa.py:1005-1006`).

On the registered `control_mode="horde_ac"` path it is dropped on the floor.
`alberta_framework/pipeline.py:1576-1581` calls

```python
ac_result = ac.update(
    ac_state,
    reward,
    features,
    auxiliary_cumulants=auxiliary_cumulants,
)
```

with no `discount`, so `HordeActorCriticAgent.update` falls back to
`value_discount = self._critic.horde_spec.gammas[cfg.value_head_index]`
(`alberta_framework/core/horde_actor_critic.py:1001-1003`) on *every* step,
terminal or not.

Two things are consequently wrong at a terminal transition:

1. **The value head bootstraps through termination.** The critic's TD target
   for the value demon is `r + gamma * V(s')` instead of `r`.
2. **The actor eligibility trace survives the boundary.** In
   `HordeActorCriticAgent.update`, `actor_decay = value_discount * cfg.actor_lamda`
   and `carry_traces = value_discount != 0.0` (`horde_actor_critic.py:1044,1061`),
   so with `discount = 0` the actor trace is zeroed — but the pipeline never
   supplies it, and the trace is carried into the next episode instead.

The mechanism to fix it already exists and was built for exactly this caller.
`HordeActorCriticAgent.update` takes `discount: Array | None` — "Optional
scalar per-transition value-head discount" — and routes to
`HordeLearner.update_with_discounts`, whose docstring says it "lets control
adapters zero the value head at episode boundaries while keeping the Horde's
fixed GVF metadata and per-head trace decay intact"
(`alberta_framework/core/horde.py:554-557`). The pipeline is such a control
adapter and is the only in-tree one that does not use it.

## Reproduction

`alberta_framework/pipeline.py` at `c9aba7b5`, identity Step 2 features,
`horde.gammas=(0.9, 0.5)`, `value_head_index=0`, `actor_lamda=0.8`:

```python
state = pipeline.init(jr.key(0), jnp.asarray([0.2, -0.1, 0.4], dtype=jnp.float32))
obs = jnp.asarray([0.1, 0.3, -0.2], dtype=jnp.float32)
reward = jnp.asarray(0.5, dtype=jnp.float32)
cumulants = jnp.asarray([0.5, -0.2], dtype=jnp.float32)
for term in (0.0, 1.0):
    res = pipeline.update(state, obs, reward, jnp.asarray(term, dtype=jnp.float32), cumulants)
    print(term, res.horde_td_targets, res.state.control_state.actor_trace_weights)
```

```
0.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
1.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
```

The two lines are byte-identical: `terminated` has **no effect whatsoever** on
this control path. The value-head target should be `0.5` (the reward alone) in
the terminal case; it is `0.5 + 0.9 * V(s') = 0.5 + 0.9 * (-0.1718) = 0.3454`.
The actor trace should be empty; it is not.

## Why no test caught it

Every existing pipeline test passes `terminated = 0.0`
(`tests/test_pipeline.py:237`, `:280`, `:324`, `:412`, `:484`, `:540`, `:580`;
the array-run helper builds `terminated = jnp.zeros(n_steps)` at `:632`). `terminated = 1.0` is exercised nowhere in
the pipeline suite, on either control mode, so the entire termination semantics
of the documented flag are untested and the `horde_ac` branch's untested path
is silently wrong.

## Scope

`AlbertaPipeline.run_arrays` delegates to `AlbertaPipeline.update`, so array
runs and `run_pipeline_smoke` inherit the same defect. Any measurement taken
through the pipeline with `control_mode="horde_ac"` on an episodic stream
learns terminal transitions as if the episode continued. Auxiliary GVF demons
are *not* affected and must keep their own configured gammas — only the
value head's discount is a per-transition control quantity.

## Proposed fix

Pass the terminal-aware discount in the `horde_ac` branch:

```python
value_gamma = self._horde.horde_spec.gammas[value_index]
control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))
ac_result = ac.update(..., discount=control_discount)
```

At `terminated == 0` this is a bit-exact reduction: `update_with_discounts`
computes `bootstrap = jnp.where(discounts == 0.0, 0.0, discounts * next_preds)`
from the same `gammas` vector the `discount is None` branch uses, and the extra
finite/range head-validity conditions are satisfied by the configured gammas.

Lane/metric: this is a measurement-trustworthiness fix, not a benchmark climb —
no `outputs/` artifact changes and no registered evidence source is touched.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2344.

## Summary
On the `control_mode="horde_ac"` path in `AlbertaPipeline.update`, `terminated` was validated but dropped when calling `HordeActorCriticAgent.update`. Consequently, at terminal episode transitions (`terminated == 1.0`), the value-head demon continued bootstrapping on the next state (`r + gamma * V(s')` instead of `r`), and the actor trace survived across the episode boundary.

## Proposed Changes
1. Computed `control_discount = jnp.where(terminated != 0.0, jnp.zeros_like(value_gamma_arr), value_gamma_arr)` in `AlbertaPipeline.update` on the `horde_ac` branch.
2. Passed `discount=control_discount` to `HordeActorCriticAgent.update`.
3. Added unit tests in `tests/test_pipeline.py` verifying that `terminated=1.0` zeroes value-head bootstrapping.

## Test Plan
- `uv run pytest tests/test_pipeline.py` (184 passed)
- `uv run ruff check alberta_framework/pipeline.py tests/test_pipeline.py` (clean)
- `uv run mypy alberta_framework/pipeline.py` (clean)

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
**APPROVE** — verified by independent sandbox reproduction at head `f5f1b597`, base `c9aba7b5`.

## Scope
Fixes #2344. Tightly scoped: the `control_mode == "horde_ac"` branch of `AlbertaPipeline.update` (pipeline.py:1576–1590) now computes a terminal-aware `control_discount` and threads it as `discount=` into `HordeActorCriticAgent.update`; plus one regression test. No other control modes or unrelated behavior touched.

## What I verified (not just read)

Toolchain note: the package needs Python 3.12 (PEP 695 generics in `core/learners.py:518`), so I provisioned a standalone CPython 3.12.7 + `jax 0.11.1`/`numpy 2.5.2` and exercised the real `AlbertaPipeline` via `_small_horde_ac_config`.

**1. Both claimed defects reproduced on base.** Same transition applied with `terminated=0.0` vs `terminated=1.0`:

```
############ BASE (c9aba7b5) ############
value_td_target  non-terminal=0.33678791  terminal=0.33678791   (reward alone = 0.5)
actor_trace |sum| non-terminal=0.700000   terminal=0.700000
```

At the terminal step the value head still bootstraps (target `= r + gamma*V(s') = 0.337`, identical to the non-terminal target instead of `r = 0.5`), and the actor eligibility trace (`|sum| = 0.70`) survives across the episode boundary. This confirms the issue: `terminated` was validated but dropped before `ac.update`.

**2. Head fixes both:**

```
############ HEAD (f5f1b597) ############
value_td_target  non-terminal=0.33678794  terminal=0.5
actor_trace |sum| non-terminal=0.700000   terminal=0.000000
```

Terminal TD target is now `r` alone and the actor trace is zeroed at the boundary. I confirmed this is not merely the diff: `HordeActorCriticAgent.update` (horde_actor_critic.py:1001–1073) threads the discount into **both** the value-head critic (`update_with_discounts`, value gamma → 0 at terminal) **and** the actor path (`actor_decay = value_discount * actor_lamda`; `carry_traces = value_discount != 0.0` resets `actor_trace_weights/bias`). So the description's dual claim is genuinely satisfied, not just the value head.

**3. Non-terminal behavior preserved (regression check).** The fix now always routes the non-terminal path through `update_with_discounts` instead of `update`. Base vs head non-terminal value target: `0.33678791` vs `0.33678794` — a ~3e-8 difference at float32-ULP level (float32 eps ≈ 1.2e-7), i.e. a benign reassociation, not a semantic change. The full touched-module suite tolerates it: `pytest tests/test_pipeline.py` → **184 passed** on head.

## Non-blocking note
Given this repo's emphasis on exact/deterministic replay, be aware that the non-terminal `horde_ac` value target shifts by ~1 ULP because it now goes through `update_with_discounts(gammas.at[value_index].set(value_gamma))` rather than the former `update(...)`. No current golden test pins it (184/184 green), so this is informational only — if bit-exact horde_ac replay fixtures are added later they should be generated against this path.

Correct, minimal, well-tested fix that resolves both terminal-transition defects with no semantic regression on the non-terminal path. No correctness, security, or test-coverage gap found.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@9b93ec4374878428a7a53b24202dfa15a8dbb55b:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Reviewed: Superseded by #2348: same terminal discount correction with stronger runtime coverage. Its regression fails on main and all 184 pipeline tests pass locally; required CI is running.
